### ISSUE TITLE
Different package name makes the different result

### ISSUE THREAD

**Issue by boringwork** [golang/go#25868]
Different package name makes the different result, when I use package `app`, the result is 
```
Student study
Student study
```

when I change the package name from app to stu, the result is 
```
BoyStudent study
GirlStudent study
```
I want the second result but unaffected for the package name

![](https://user-images.githubusercontent.com/4026598/41359823-e92a4f22-6f5d-11e8-81f7-6951dc903da0.png)

```go
package app

import (
	"fmt"
)

type IStudent interface {
	study()
}

type Student struct {
}

func (me *Student) study() {
	fmt.Println("Student study")
}

type StudentProxy struct {
	student IStudent
}

func (me *StudentProxy) study() {
	me.student.study()
}

type studentManager struct {
	students map[string]*StudentProxy
	finder   func(string) IStudent
}

var sm = &studentManager{
	students: map[string]*StudentProxy{},
}

func StudentManager() *studentManager {
	return sm
}

func (me *studentManager) SetFinder(finder func(string) IStudent) {
	me.finder = finder
}

func (me *studentManager) GetStudent(name string) *StudentProxy {
	proxy, ok := me.students[name]
	if !ok {
		stu := me.finder(name)
		proxy = &StudentProxy{student: stu}
		me.students[name] = proxy
	}
	return proxy
}

func (me *studentManager) MethodA() {
	names := []string{"boy", "girl"}
	for _, name := range names {
		me.GetStudent(name).study()
	}
}

func (me *studentManager) MethodB() {
	names := []string{"boy", "girl"}
	for _, name := range names {
		me.GetStudent(name).study()
	}
}
```

```go
package auto

import (
	"github.com/boringwork/godemo/interface/app"
	"github.com/boringwork/godemo/interface/impl"
)

func Finder(name string) app.IStudent {
	switch name {
	case "boy":
		return &impl.BoyStudent{}
	case "girl":
		return &impl.GirlStudent{}
	}
	return nil
}
```

```go
package impl

import (
	"fmt"

	"github.com/boringwork/godemo/interface/app"
)

type BoyStudent struct {
	app.Student
}

func (me *BoyStudent) study() {
	fmt.Println("BoyStudent study")
}

type GirlStudent struct {
	app.Student
}

func (me *GirlStudent) study() {
	fmt.Println("GirlStudent study")
}
```

```go
package main

import (
	"github.com/boringwork/godemo/interface/app"
	"github.com/boringwork/godemo/interface/auto"
)

func main() {
	app.StudentManager().SetFinder(auto.Finder)
	app.StudentManager().MethodA()
}
```

### What version of Go are you using (`go version`)?
go version go1.10.1 darwin/amd64

**Comment by bradfitz** [comment]
This doesn't look like a bug report but rather a question about the language. For questions about Go, see https://golang.org/wiki/Questions.

**Comment by boringwork** [comment]
The package name affecte result of program, why did you think this is not a bug?

**Comment by AlexRouSg** [comment]
@bradfitz 
I find this weird too, simplified repo below.
Not sure which's the correct behavior as this is about an unexported function but it should either print
```
B
B
```
or
```
A
C
```

But it's printing 
```
A
B
```

```
package main

import (
	"a"
	"b"
	"c"
)

func main() {
	testA := a.A{}
	testC := c.C{}
	b.Test(testA)
	b.Test(testC)
}
```

```
package a

import (
	"fmt"

	"b"
)

type A struct {
	b.B
}

func (A) foo() {
	fmt.Println("A")
}
```

```
package b

import (
	"fmt"
)

type Foo interface {
	foo()
}

type B struct {
}

func (B) foo() {
	fmt.Println("B")
}

func Test(f Foo) {
	f.foo()
}
```

```
package c

import (
	"fmt"

	"c"
)

type C struct {
	b.B
}

func (C) foo() {
	fmt.Println("C")
}
```

**Comment by ianlancetaylor** [comment]
@AlexRouSg I can't recreate that.  For me it prints

```
B
B
```

which is what I would expect.  How exactly did you build the program?

I did this:

```
> ls -R /tmp/x/src
.:
a/  b/	c/  main/

./a:
a.go

./b:
b.go

./c:
c.go

./main:
main.go
> cd /tmp/x/src/main
> GOPATH=/tmp/x go build
> ./main
B
B
```

**Comment by AlexRouSg** [comment]
@ianlancetaylor I did a` go run main.go` on 1.10.3

**Comment by ianlancetaylor** [comment]
@AlexRouSg Is the layout of files exactly as I show above?  What is `GOPATH` set to?

**Comment by ianlancetaylor** [comment]
Ah, I see that with Go 1.10 I do get

```
A
B
```

This may be #24693, which is fixed in 1.11.

**Comment by AlexRouSg** [comment]
Yup on tip I get 
```
B
B
```
